### ISSUE TITLE
Random Seed / Calling GMM-Demux from python code?

### ISSUE THREAD

**Issue by bbimber** [CHPGenetics/GMM-demux#4]
Hello,

I have an R package that runs various cellhashing algorithms and compares the results. I'd like to incorporate gmm-demux, but have two questions:

1) Is there a way to call gmm-demux from python code (as opposed to executing on the command line)? I think I can in theory call GMM_Demux.main() directly, but when I tried this it was failing with arg-parsing errors. 

2) Is there a way to set the random seed used by gmm-demux? the results do not seem deterministic, and this is my only guess as to why.

**Comment by xhongyi** [comment]
Hi bbimber,

Sorry for the late reply. Let me look into this. I will post a follow up by 8/2.

Hongyi

**Comment by xhongyi** [comment]
It seems that converting GMM-Demux to a library requires some work. Perhaps it will be added in future releases.

At the moment, the best option, perhaps, would be DIY modifications to GMM-Demux.

For your first question, I would recommend you take a look at the GMM-Demux.py file line 54-101. You may want to extract the ["Cluster_id"] column from the GMM_full_df collection.

Now for your second question, the randomness of GMM-Demux should be minor, assuming that the distributions of the data follow moderately-good Gaussian distributions.

Nevertheless, the randomness stems from line 23 in classify_drops.py.

To remove randomness, set the random_state parameter of the GaussianMixture class to a fixed value. Please refer to the sklearn documentation for additional information:

https://scikit-learn.org/stable/modules/generated/sklearn.mixture.GaussianMixture.html

**Comment by bbimber** [comment]
@xhongyi Thanks for the reply. I ended up wrapping GMM-demux like this:

https://github.com/BimberLab/cellhashR/blob/a1d5b91d69f18a736bb539953826908dd50c5b25/R/GMM_Demux.R#L24

this lets our cellhashR package run several cell hashing algorithms in parallel on data and give summary results.

I will check into the random seed info you just sent.

**Comment by bbimber** [comment]
@xhongyi to the point about random seed: would you consider supporting a command line argument to provide the random seed? it seems like it would be really easy to pass this, using obtain_arrays(data, seed), which would pass it to GaussianMixture()? So long as argument had a reasonable default value this would be pretty non-invasive?

**Comment by xhongyi** [comment]
That can definitely be done!

Let me add that option by this weekend!

**Comment by bbimber** [comment]
@xhongyi OK, great. As you probably saw I took a stab at this with the PR here: #5.

**Comment by bbimber** [comment]
@xhongyi hey, just wanted to follow up. what do you think about the solution I proposed in this PR: #5 ?

**Comment by bbimber** [comment]
@xhongyi Hello - I just thought I'd check back one last time on whether you're willing to accept PR #5. As far as we can see, setting random-state is now giving us a deterministic output from gmm-demux. We're building off my fork; however, I'd love to see this integrated into the primary tool so others can take advantage of this.

**Comment by xhongyi** [comment]
Hi bbimber,

Yeah I will look into this. I am only the author of the code but not the owner of the repo.

I will let the owner know.

Thanks,
Hongyi

**Comment by bbimber** [comment]
@xhongyi OK, I appreciate that. We've been running our fork of this repo (with the random seed addition), and it seems to be working fine.

**Comment by xhongyi** [comment]
Hi bbimber,

I've gained admin control of the repo and I have approved your merge request.

Thanks for your contribution!! Much appreciated.

Thanks,
Hongyi

**Comment by bbimber** [comment]
@xhongyi wow, i assumed this PR was dead. thanks for finally merging it!
